openssl/src/ts.rs

```diff
@@ -91,8 +92,13 @@ impl TsMsgImprint {
     ///
     /// `hash` must have originated from the hash function specified by `md`.
     pub fn from_prehash_with_algo(hash: &[u8], md: MessageDigest) -> Result<Self, ErrorStack> {
-        let mut algo = X509Algorithm::new()?;
-        algo.set_md(md);
+        let algo = X509Algorithm::new()?;
+
+        let aobj = unsafe { cvt_p(OBJ_nid2obj(EVP_MD_get_type(md.as_ptr())))? };
+        let res = unsafe { X509_ALGOR_set0(algo.as_ptr(), aobj, ffi::V_ASN1_NULL, ptr::null_mut()) };
+        cvt(res).inspect_err(|_| unsafe {
+            ASN1_OBJECT_free(aobj);
```

Comment:
Thanks, this part looks good now.

This free isn't really needed, but doesn't hurt (OBJ_nid2obj() returns objects from an internal hash, the result never needs to be freed)..